src/Common/Auth/AuthUtils.php

```diff
@@ -0,0 +1,548 @@
+     * @param type $currentPwd      the active user's current password - CAN NOT BE EMPTY
+     *                              - password is passed by reference so that it can be "cleared out" as soon as we are done with it.
+     * @param type $newPwd          the new password for the target user
+     *                              - password is passed by reference so that it can be "cleared out" as soon as we are done with it.
+     * @param type $errMsg          passed by reference to return any
+     * @param type $create          Are we creating a new user or
+     * @param type $insert_sql      SQL to run to create the row in "users" (and generate a new id) when needed.
+     * @param type $new_username    The username for a new user
+     * @param type $newid           Return by reference of the ID of a created user
+     * @return boolean              Was the password successfully updated/created? If false, then $errMsg will tell you why it failed.
+     */
+    public static function updatePassword($activeUser, $targetUser, &$currentPwd, &$newPwd, &$errMsg, $create = false, $insert_sql = "", $new_username = null, &$newid = null)
+    {
+        if (empty($activeUser) || empty($currentPwd)) {
+            $errMsg = xl("Password update error!");
+            self::clearFromMemory($currentPwd);
+            self::clearFromMemory($newPwd);
+            return false;
+        }
+
+        $userSQL = "SELECT `password`, `password_history1`, `password_history2`" .
+            " FROM `users_secure`" .
+            " WHERE `id` = ?";
+        $userInfo = privQuery($userSQL, [$targetUser]);
+
+        // Verify the active user's password
+        $changingOwnPassword = $activeUser==$targetUser;
+        // True if this is the current user changing their own password
+        if ($changingOwnPassword) {
+            if ($create) {
+                $errMsg = xl("Trying to create user with existing username!");
+                self::clearFromMemory($currentPwd);
+                self::clearFromMemory($newPwd);
+                return false;
+            }
+            if (empty($userInfo['password'])) {
+                $errMsg = xl("Password update error!");
+                self::clearFromMemory($currentPwd);
+                self::clearFromMemory($newPwd);
+                return false;
+            }
+            // If this user is changing his own password, then confirm that they have the current password correct
+            if (!password_verify($currentPwd, $userInfo['password'])) {
+                $errMsg = xl("Incorrect password!");
+                self::clearFromMemory($currentPwd);
+                self::clearFromMemory($newPwd);
+                return false;
+            }
+        } else {
+            // If this is an administrator changing someone else's password, then check that they have the password right
+            if (self::useActiveDirectory()) {
```

Comment:
Clarify with comment that this active directory code is only in the case to verify the admin user when adding a new user (and the random password is then used for the new user)